sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java

```diff
@@ -149,6 +158,64 @@ private void createMetaDataBasedIterables(MetaData metaData) {
         ingestionRules = new IngestionRuleInfos(metaData.custom(IngestRulesMetaData.TYPE));
     }
 
+    /**
+     * Iterable for extracting not null constraints from table info.
+     */
+    static class NotNullConstraintIterable implements Iterable<ConstraintInfo> {
+
+        private final TableInfo ti;
+
+        NotNullConstraintIterable(TableInfo ti) {
+            this.ti = ti;
+        }
+
+        @Override
+        public Iterator<ConstraintInfo> iterator() {
+            return new NotNullConstraintIterator(ti);
+        }
+    }
+
+    /**
+     * Iterator that returns ConstraintInfo for each TableInfo with not null constraints.
+     */
+    static class NotNullConstraintIterator implements Iterator<ConstraintInfo> {
+        private final TableInfo tableInfo;
+        private final Iterator<Reference> nullableColumns;
+
+        NotNullConstraintIterator(TableInfo tableInfo) {
+            this.tableInfo = tableInfo;
+            nullableColumns = StreamSupport.stream(tableInfo.spliterator(), false)
+                .filter(reference -> !reference.ident().columnIdent().isSystemColumn()
+                                     && reference.valueType() != DataTypes.NOT_SUPPORTED
+                                     && !reference.isNullable()).iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return nullableColumns.hasNext();
+        }
+
+        @Override
+        public ConstraintInfo next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException("Primary Key iterator exhausted");
+            }
+
+            // Return nullable columns instead.
+            return new ConstraintInfo(
+                this.tableInfo.ident(),
+                String.format(
```

Comment:
Using a StringBuilder instead of `String.format` would be more efficient here. Given that it's done per row I think this would be worth changing.